add-custom-dynamic-entity-data-source/src/main/java/com/esri/arcgismaps/sample/addcustomdynamicentitydatasource/components/CustomEntityFeedProvider.kt

```diff
@@ -0,0 +1,172 @@
+import kotlinx.serialization.json.contentOrNull
+import kotlinx.serialization.json.double
+import kotlinx.serialization.json.jsonObject
+import kotlinx.serialization.json.jsonPrimitive
+import java.io.File
+import java.io.IOException
+
+class CustomEntityFeedProvider(
+    fileName: String,
+    private val entityIdField: String,
+    private val delayDuration: Long
+) :
+    CustomDynamicEntityDataSource.EntityFeedProvider {
+
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    private val observationsFile = File(fileName)
+
+    // Create a shared flow to emit feed events.
+    private val _feed = MutableSharedFlow<CustomDynamicEntityDataSource.FeedEvent>(
+        extraBufferCapacity = Int.MAX_VALUE,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+
+    // Expose the feed as a shared flow.
+    override val feed = _feed
+
+    // Keep track of the feed job to allow us to properly cancel it when needed.
+    private var feedJob: Job? = null
+
+    /**
+     * Called when the data source is connected. Checks for presence of the observations file and
+     * starts reading the file asynchronously.
+     */
+    override suspend fun onConnect() {
+        if (!observationsFile.exists()) {
+            throw IOException("Observations file does not exist.")
+        }
+
+        readObservationsFileAsync()
+    }
+
+    /**
+     * Called when the data source is disconnected. Cancels all child coroutines.
+     */
+    override suspend fun onDisconnect() {
+        feedJob?.cancelAndJoin()
+        feedJob = null
+    }
+
+    /**
+     * Called when the data source is loaded. Defines the Dynamic Entity Data Source info.
+     */
+    override suspend fun onLoad(): DynamicEntityDataSourceInfo {
+        return DynamicEntityDataSourceInfo(entityIdField, getSchema()).apply {
+            spatialReference = SpatialReference.wgs84()
+        }
+    }
+
+    /**
+     * Reads the observations file asynchronously and emits feed events for each observation.
+     */
+    private fun readObservationsFileAsync() {
+        feedJob = scope.launch(Dispatchers.IO) {
+            try {
+                // While no call to cancel the job has been made.
+                observationsFile.bufferedReader().use { reader ->
+                    // Read the next line from the file.
+                    for (line in reader.lines()) {
+                        // Adjusting the value for the delay will change the speed at which the
+                        // entities and their observations are displayed.
+                        delay(delayDuration)
+                        // Emit the next observation.
+                        _feed.tryEmit(processNextObservation(line))
+                    }
+                }
+            } catch (e: Exception) {
+                if (e is CancellationException) {
+                    throw e
+                }
+                withContext(NonCancellable) {
+                    _feed.tryEmit(
```

Comment:
I suggest the following comment:
```suggestion
                    // Signal that an error occured. This will change the CustomDynamicEntityDataSource's connectionStatus to Failed.
                    _feed.tryEmit(
```